pkg/services/sqlstore/migrations/annotation_mig.go

```diff
@@ -109,4 +110,51 @@ func addAnnotationMig(mg *Migrator) {
 	//
 	updateEpochSql := "UPDATE annotation SET epoch = (epoch*1000) where epoch < 9999999999"
 	mg.AddMigration("Convert existing annotations from seconds to milliseconds", NewRawSqlMigration(updateEpochSql))
+
+	//
+	// 6.3: Make Regions a single annotation row
+	//
+	mg.AddMigration("Add epoch_end column", NewAddColumnMigration(table, &Column{
+		Name: "epoch_end", Type: DB_BigInt, Nullable: false, Default: "0",
+	}))
+	mg.AddMigration("Add index for epoch_end", NewAddIndexMigration(table, &Index{
+		Cols: []string{"epoch_end"}, Type: IndexType,
+	}))
+	mg.AddMigration("Make epoch_end the same as epoch", NewRawSqlMigration("UPDATE annotation SET epoch_end = epoch"))
+	mg.AddMigration("Move region to single row", &AddMakeRegionSingleRowMigration{})
+
+	// TODO! drop region_id column?
+}
+
+type AddMakeRegionSingleRowMigration struct {
+	MigrationBase
+}
+
+func (m *AddMakeRegionSingleRowMigration) Sql(dialect Dialect) string {
+	return "code migration"
+}
+
+type TempRegionInfoDTO struct {
+	RegionId int64
+	Epoch    int64
+}
+
+func (m *AddMakeRegionSingleRowMigration) Exec(sess *xorm.Session, mg *Migrator) error {
+	regions := make([]*TempRegionInfoDTO, 0)
```

Comment:
Maybe we should consider using a transaction? Not sure what possible consequences that may have...

Reply:
I thought each migration already happens in a transaction?  See:

https://github.com/grafana/grafana/blob/master/pkg/services/sqlstore/migrator/migrator.go#L97

```
		err := mg.inTransaction(func(sess *xorm.Session) error {
			err := mg.exec(m, sess)
                         ...
```

Reply:
Yeah that's true :+1: